File: api/v1/schemas/admin_users.py
# -*- coding: utf-8 -*-
"""Admin user management schemas."""

from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class AdminUserSummary(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: int
    username: str
    email: Optional[str] = None
    role: str
    status: str
    created_at: Optional[str] = Field(default=None, alias="createdAt")
    updated_at: Optional[str] = Field(default=None, alias="updatedAt")
    last_login_at: Optional[str] = Field(default=None, alias="lastLoginAt")


class AdminUserListResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    items: List[AdminUserSummary]
    total: int
    page: int
    page_size: int = Field(alias="pageSize")


class CreateAdminUserRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    username: str = Field(min_length=1, max_length=64)
    password: str = Field(min_length=6, max_length=128)
    role: Literal["admin", "user"] = "user"
    email: Optional[str] = None


class UpdateAdminUserRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    role: Optional[Literal["admin", "user"]] = None
    status: Optional[Literal["active", "disabled"]] = None
    email: Optional[str] = None


class ResetAdminUserPasswordRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    new_password: str = Field(alias="newPassword", min_length=6, max_length=128)

File: apps/dsa-web/src/api/adminUsers.ts
import apiClient from './index';

export type AdminUserSummary = {
  id: number;
  username: string;
  email?: string | null;
  role: 'admin' | 'user' | string;
  status: 'active' | 'disabled' | string;
  createdAt?: string | null;
  updatedAt?: string | null;
  lastLoginAt?: string | null;
};

export type AdminUserListResponse = {
  items: AdminUserSummary[];
  total: number;
  page: number;
  pageSize: number;
};

export type CreateAdminUserRequest = {
  username: string;
  password: string;
  role?: 'admin' | 'user';
  email?: string | null;
};

export type UpdateAdminUserRequest = {
  role?: 'admin' | 'user';
  status?: 'active' | 'disabled';
  email?: string | null;
};

export const adminUsersApi = {
  async list(params?: {
    page?: number;
    pageSize?: number;
    search?: string;
    role?: string;
    status?: string;
  }): Promise<AdminUserListResponse> {
    const { data } = await apiClient.get<AdminUserListResponse>('/api/v1/admin/users', {
      params: {
        page: params?.page,
        pageSize: params?.pageSize,
        search: params?.search || undefined,
        role: params?.role || undefined,
        status: params?.status || undefined,
      },
    });
    return data;
  },

  async create(body: CreateAdminUserRequest): Promise<AdminUserSummary> {
    const { data } = await apiClient.post<AdminUserSummary>('/api/v1/admin/users', body);
    return data;
  },

  async update(userId: number, body: UpdateAdminUserRequest): Promise<AdminUserSummary> {
    const { data } = await apiClient.patch<AdminUserSummary>(`/api/v1/admin/users/${userId}`, body);
    return data;
  },

  async disable(userId: number): Promise<void> {
    await apiClient.delete(`/api/v1/admin/users/${userId}`);
  },

  async resetPassword(userId: number, newPassword: string): Promise<void> {
    await apiClient.post(`/api/v1/admin/users/${userId}/reset-password`, {
      newPassword,
    });
  },

  async revokeSessions(userId: number): Promise<void> {
    await apiClient.post(`/api/v1/admin/users/${userId}/revoke-sessions`);
  },
};
